### ISSUE TITLE
Integrate GitLab with Kits

### ISSUE THREAD

**Issue by joliet** [DukeLearningInnovation/kits#487]
I am a Code+ Program lead trying to use Kit for my team but GitLab is not an app I can use with Kits. This makes me frustrated because using Kits would be a good fit for my team's collaboration needs.

**Questions**

**Assumptions**

**Hypotheses**

**Comment by joliet** [comment]
Use cases:

Single repository
Multiple repositories
Repository for all group members
Look at roles

**Comment by profmikegreene** [comment]
It'd be interesting to review Github Classroom and see how much of those workflows could be replicated in Gitlab. That may be out of scope and something more suitable for Drew to work on, not sure.

**Comment by joliet** [comment]
GitLab API documentation - https://docs.gitlab.com/ee/api/

**Comment by joliet** [comment]
Feature doc for this integration: https://docs.google.com/document/d/1Di4xzlej4Am3wc_krW6R_lqP8Eq330HDFB4h2X76J7s/edit

**Comment by joliet** [comment]
Hi-level workflow and mapping roles between Kits and GitLab - https://docs.google.com/spreadsheets/d/1PT7uJaX3WgYikoSsYWXh8qfbiqeOQURu7IumZX4gWII/edit#gid=1142634128

Will include links to Figma mockups for UI design for this integration.

**Comment by ewendland23** [comment]
Hi @joliet 

GitLab is available now on Kits-test for you to start playing with.

I have removed you from gitlab-dev so it will need to create a new GitLab account for you if you add the Git App to a kit.  I figured out how to prevent it from sending an email that you need to use to log in.

I'm out next week, but if you wanted to start testing, feel free.

Thanks and Happy Holiday!
Liz

**Comment by joliet** [comment]
GitLab feature narrative - https://docs.google.com/document/d/1Di4xzlej4Am3wc_krW6R_lqP8Eq330HDFB4h2X76J7s/edit

**Comment by joliet** [comment]
Thanks @ewendland23 .  This looks like a really good start.  I have the app page content 90% ready for the app store.

Here is the link to the mockup in Figma for the GitLab integration - https://www.figma.com/file/qXbuKueTY3tShyQxeG3t70/App-Integrations?node-id=0%3A2265
(GitLab mockups at bottom)

This includes the options for choosing a different mapping for the student role in course kits and the viewer role in collaboration kits.  I added the 'Learn more about roles' option we talked about that has a link through to GitLab's documentation on roles.

I updated this doc and noticed that we have decided to make the group title editable. https://docs.google.com/spreadsheets/d/1PT7uJaX3WgYikoSsYWXh8qfbiqeOQURu7IumZX4gWII/edit?usp=sharing
 I don't think we do this with other apps (MS Teams, Panopto, etc).  Was this a mistake or did we discuss allowing the group title to be edited?

**Comment by ewendland23** [comment]
This is ready on Kits-test.  We do not let users change the name of the GitLab Group.  We do allow them to set the role for students/editors and viewers.  We currently are NOT creating a project.

**Comment by joliet** [comment]
@noahhh What do you need for the App Store card for GitLab in terms of an image?  I uploaded something to Drupal but it's not showing up.  Happy to provide that here so we can get that piece updated.

**Comment by noahhh** [comment]
@joliet in the past an .svg has been provided that I can add to the css for this

**Comment by joliet** [comment]
Thanks @noahhh .  Here are the svgs for both the kit icon and the app store card.
![GitLab app](https://user-images.githubusercontent.com/7387760/151807406-e2f2a1c6-2e20-4686-b821-e591f41c061d.svg)
![GitLab logo](https://user-images.githubusercontent.com/7387760/151807408-6f38faca-e455-40b4-bcb8-d086f251c27a.svg)
.

**Comment by noahhh** [comment]
@joliet thanks, this is out in kits test now

**Comment by joliet** [comment]
Hi @ewendland23  I heard from our GitLab users and none need a project set up as part of the integration.  However, one person asked if we could include a remapping option for the TA role.  Thinking about how we might add that, I realized that our set up page for a course has both the Student and Viewer role options to re-map.  That might be confusing.  We should either make those options contextual (student on course kits and viewer on collaboration kits) or add some text to make this more clear.

**Comment by ewendland23** [comment]
@joliet, I am not sure I am following when you say 'make those options contextual'.  We already are making it contextual.  For 'Course-type' kits we show student and viewer.  For Collaboration kits we show editor and viewer.  I am guessing for Course we could also show a TA dropdown?  But would that also change Assistant's role?

**Comment by joliet** [comment]
Hi @ewendland23 .  There isn't a viewer role in a course kit.  Student is equivalent to Viewer in collaboration kits.  In Kits-test when I add GitLab to an official course kit, I see two drop-down options, one for Student and one for Viewer.  The Viewer option might be confusing in this case.  This might be my fault because I only gave you one mockup that showed both options - Student drop-down options for course kits and Viewer drop-down options for collaboration kits.  I made separate mockups and included the Assistant drop-down options for course kits and the Editor drop-down options for collaboration kits - https://www.figma.com/file/qXbuKueTY3tShyQxeG3t70/App-Integrations?node-id=0%3A2265

**Comment by ewendland23** [comment]
Hi @joliet . I believe I have the GitLab app acting like you want now.  Please test and let me know if you see any issues.

**Comment by joliet** [comment]
This is working as expected @ewendland23 .  If could make a few small UI changes before we move it to prod, that would be fantastic.  These are - change the label for the title to "Group title:"  We won't be creating a project as part of the setup.  The second tweak is around the alignment of the elements.  Everything is pushed too far left.  Here are some examples comparing the second GitLab setup page to the second page for Panopto.

<img width="1661" alt="GitLab_Setup" src="https://user-images.githubusercontent.com/7387760/152434582-a052ff7f-8c4a-4f41-854a-3e5fc2e376ce.png">

<img width="1669" alt="Panopto_Setup" src="https://user-images.githubusercontent.com/7387760/152434609-034e8a48-f6a3-4dd8-986b-2b43fc45fba5.png">

**Comment by ewendland23** [comment]
Hi @joliet 
These changes are now on Kits Test.  Thanks for the excellent description of what you wanted!

**Comment by joliet** [comment]
Hi @ewendland23 .  Thanks for these updates.  When I tested with kit id 3013 on kits-test, I saw a weird character in the group title.  Here's a screenshot.

<img width="490" alt="Screen Shot 2022-02-08 at 9 46 25 AM" src="https://user-images.githubusercontent.com/7387760/153011070-25a61f1d-06f8-4120-835b-acd9609547b3.png">

If we can figure that out, I think this is ready for prod.  I would like to do just a little more testing today before sign off, just to be sure.
